cmd/list.go

```diff
@@ -0,0 +1,280 @@
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+	"text/tabwriter"
+
+	"github.com/urfave/cli"
+
+	"github.com/manifoldco/torus-cli/api"
+	"github.com/manifoldco/torus-cli/apitypes"
+	"github.com/manifoldco/torus-cli/config"
+	"github.com/manifoldco/torus-cli/envelope"
+	"github.com/manifoldco/torus-cli/errs"
+	"github.com/manifoldco/torus-cli/pathexp"
+)
+
+func init() {
+	list := cli.Command{
+		Name:      "list",
+		ArgsUsage: "",
+		Usage:     "List your org's project structure and its secrets.",
+		Category:  "SECRETS",
+		Flags: []cli.Flag{
+			orgFlag("Use this organization.", false),
+			projectFlag("Use this project.", false),
+			envSliceFlag("Use this environment.", false),
+			serviceSliceFlag("Use this service.", "", false),
+			nameFlag("Find secrets with this name."),
+			cli.BoolFlag{
+				Name:  "verbose, v",
+				Usage: "Lists the sources of the secrets (shortcut for --format verbose)",
+			},
+		},
+		Action: chain(
+			ensureDaemon, ensureSession, loadDirPrefs, checkRequiredFlags, listCmd,
+		),
+	}
+	Cmds = append(Cmds, list)
+}
+
+func listCmd(ctx *cli.Context) error {
+	verbose := ctx.Bool("verbose")
+
+	args := ctx.Args()
+
+	cfg, err := config.LoadConfig()
+	if err != nil {
+		return err
+	}
+
+	client := api.NewClient(cfg)
+	c := context.Background()
+
+	// TODO - Make this a type
+	credentialsTree := make(map[string]map[string]credentialSet)
+
+	// Retrieve org and project flag values
+	orgName := ctx.String("org")
+	var org *envelope.Org
+	if orgName == "" {
+		// Retrieve list of available orgs
+		orgs, err := client.Orgs.List(c)
+		if err != nil {
+			return errs.NewExitError("Failed to retrieve orgs list.")
+		}
+
+		// Prompt user to select from list of existing orgs
+		idx, _, err := SelectExistingOrgPrompt(orgs)
+		if err != nil {
+			return errs.NewExitError("Failed to select org.")
+		}
+
+		org = &orgs[idx]
+		orgName = org.Body.Name
+
+	} else {
+		// If org flag was used, identify the org supplied.
+		org, err = client.Orgs.GetByName(c, orgName)
+		if org == nil || err != nil {
+			return errs.NewExitError("org" + orgName + "not found.")
+		}
+	}
+
+	projectName := ctx.String("project")
+	var project *envelope.Project
+	if projectName == "" {
+		// Retrieve list of available projects
+		projects, err := client.Projects.List(c, org.ID)
+		if err != nil {
+			return errs.NewExitError("Failed to retrieve projects list.")
```

Comment:
We shoudl probably use `NewErrorExitError` here and in other places so we actually return the underlying error, we don't know why it failed!

Reply:
Good catch, missed that.